src/server/frontend_wayland/wl_keyboard.cpp

```diff
@@ -69,113 +66,92 @@ mf::WlKeyboard::WlKeyboard(
 {
-    executor->spawn(run_unless(
-        destroyed,
-        [
-            ev = mir::client::Event{mir_event_ref(mir_input_event_get_event(event))},
-            this
-        ] ()
+    auto const input_ev = mir_keyboard_event_input_event(key_event);
+    auto const ev = mir::client::Event{mir_event_ref(mir_input_event_get_event(input_ev))};
+    auto const serial = wl_display_next_serial(wl_client_get_display(client));
+    auto const scancode = mir_keyboard_event_scan_code(key_event);
+    /*
+        * HACK! Maintain our own XKB state, so we can serialise it for
+        * wl_keyboard_send_modifiers
+        */
+
+    switch (mir_keyboard_event_action(key_event))
+    {
+        case mir_keyboard_action_up:
+            xkb_state_update_key(state.get(), scancode + 8, XKB_KEY_UP);
+            wl_keyboard_send_key(resource,
+                serial,
+                mir_input_event_get_event_time_ms(input_ev),
+                mir_keyboard_event_scan_code(key_event),
+                WL_KEYBOARD_KEY_STATE_RELEASED);
+            break;
+        case mir_keyboard_action_down:
+            xkb_state_update_key(state.get(), scancode + 8, XKB_KEY_DOWN);
+            wl_keyboard_send_key(resource,
+                serial,
+                mir_input_event_get_event_time_ms(input_ev),
+                mir_keyboard_event_scan_code(key_event),
+                WL_KEYBOARD_KEY_STATE_PRESSED);
+            break;
+        default:
+            break;
+    }
+    update_modifier_state();
+}
+
+void mf::WlKeyboard::handle_window_event(MirWindowEvent const* event, WlSurface* surface)
+{
+    if (mir_window_event_get_attribute(event) == mir_window_attrib_focus)
+    {
+        auto const focussed = mir_window_event_get_attribute_value(event);
+        auto const serial = wl_display_next_serial(wl_client_get_display(client));
+        if (focussed)
         {
-            auto const serial = wl_display_next_serial(wl_client_get_display(client));
-            auto const event = mir_event_get_input_event(ev);
-            auto const key_event = mir_input_event_get_keyboard_event(event);
-            auto const scancode = mir_keyboard_event_scan_code(key_event);
             /*
-                * HACK! Maintain our own XKB state, so we can serialise it for
-                * wl_keyboard_send_modifiers
+                * TODO:
+                *  *) Send the surface's keymap here.
                 */
+            auto const keyboard_state = acquire_current_keyboard_state();
 
-            switch (mir_keyboard_event_action(key_event))
+            wl_array key_state;
+            wl_array_init(&key_state);
+
+            auto* const array_storage = wl_array_add(
+                &key_state,
+                keyboard_state.size() * sizeof(decltype(keyboard_state)::value_type));
+            if (!array_storage)
+            {
+                wl_resource_post_no_memory(resource);
+                BOOST_THROW_EXCEPTION(std::bad_alloc());
+            }
```

Comment:
Throwing here risks failing to call `wl_array_release()` I guess that's pre-existing though.

Maybe it is time for a `wl_array` wrapper class?